### ISSUE TITLE
CA1859: False positive for operators and multiple derived types

### ISSUE THREAD

**Issue by ThorstenReichert** [dotnet/roslyn-analyzers#7317]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: SDK 8.0.101

### Describe the bug

CA1859 is falsely raised when multiple derivatives are involved together with custom operator overwrites. Applying the suggested fix leads to compilation errors.

### Repro

```csharp
public abstract class Base
{
    public static Base operator *(Base left, int right) => default;
}

public class A : Base;
public class B : Base;

public class Test
{
    private static Base Foo(Base input)
    //                  ^^^ CA1859
    {
        return input is A ? new B() : input * 4;
    }
}
```

### Actual behavior

CA1859 is raised for `Test.Foo` method, suggesting to change the return type to `B`

## Additional context

No warning is raised when either
- the `operator` ist replaced with a method like `static Base Multiply(Base, int)`
- the ternary operator in the return expression of `Test.Foo` is removed (e.g. `return input * 4;)`

**Comment by CollinAlpert** [comment]
Thanks, this was an excellent bug report!
